// metaapi.cloud-sdk@29.3.3/dist/helpers/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    assembleLog4jsConfig: function() {
        return assembleLog4jsConfig;
    },
    createHandlePromise: function() {
        return createHandlePromise;
    },
    delay: function() {
        return delay;
    },
    ensurePromise: function() {
        return ensurePromise;
    },
    expBackoffDelay: function() {
        return expBackoffDelay;
    },
    wait: function() {
        return wait;
    },
    waitAny: function() {
        return waitAny;
    },
    waitPass: function() {
        return waitPass;
    },
    waitTrue: function() {
        return waitTrue;
    },
    wrapHandlePromise: function() {
        return wrapHandlePromise;
    }
});
const _timeoutError = /*#__PURE__*/ _interop_require_default(require("../clients/timeoutError"));
const _logger = /*#__PURE__*/ _interop_require_default(require("../logger"));
const _lodash = /*#__PURE__*/ _interop_require_default(require("lodash"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const realSetTimeout = setTimeout;
const realDateNow = Date.now.bind(Date);
function createHandlePromise() {
    let resolve, reject;
    let promise = new Promise((res, rej)=>{
        resolve = res;
        reject = rej;
    });
    promise.completed = false;
    promise.resolve = (result)=>{
        if (!promise.completed) {
            promise.completed = true;
            promise.resolved = true;
            promise.result = result;
            resolve(result);
        }
    };
    promise.reject = (err)=>{
        if (!promise.completed) {
            promise.completed = true;
            promise.rejected = true;
            promise.error = err;
            reject(err);
        }
    };
    promise.timeout = (milliseconds, errorMessage)=>{
        if (!promise.completed) {
            let timeout = setTimeout(()=>promise.reject(new _timeoutError.default(errorMessage)), milliseconds);
            promise.finally(()=>clearTimeout(timeout)).catch(()=>{});
        }
        return promise;
    };
    promise.catch(()=>{});
    return promise;
}
function wrapHandlePromise(promise) {
    let result = createHandlePromise();
    promise.then(result.resolve).catch(()=>{});
    promise.catch(result.reject);
    return result;
}
async function ensurePromise(call) {
    return call();
}
function delay(ms, options) {
    let resolve;
    let timeout;
    let canceled = false;
    let result = new Promise((res)=>{
        timeout = (options === null || options === void 0 ? void 0 : options.ignoreSinonClock) ? realSetTimeout(res, ms) : setTimeout(res, ms);
        resolve = res;
    });
    Object.defineProperty(result, "canceled", {
        get: ()=>canceled,
        enumerable: true,
        configurable: true
    });
    result.cancel = ()=>{
        canceled = true;
        clearTimeout(timeout);
        resolve();
    };
    return result;
}
function assembleLog4jsConfig(config = {}) {
    let appenders = {
        console: {
            type: "console"
        }
    };
    let categories = {
        default: {
            appenders: Object.keys(appenders),
            level: config.defaultLevel || "INFO"
        }
    };
    Object.keys(config.levels || {}).forEach((category)=>{
        categories[category] = {
            appenders: Object.keys(appenders),
            level: config.levels[category]
        };
    });
    return {
        appenders,
        categories
    };
}
async function wait(callable, intervalInMs = 25, options) {
    const logger = _logger.default.getLogger("helpers.wait");
    if (typeof intervalInMs === "object") {
        // for backward compatibility
        options = intervalInMs;
        intervalInMs = _lodash.default.defaultTo(options.intervalInMs, 1000);
    }
    const dateNow = (options === null || options === void 0 ? void 0 : options.ignoreSinonClock) ? realDateNow : ()=>Date.now();
    let result = false, lastError;
    let timesAt = dateNow() + _lodash.default.defaultTo(options === null || options === void 0 ? void 0 : options.timeoutInMs, 30000);
    while(!result && dateNow() < timesAt){
        try {
            result = await callable();
        } catch (err) {
            lastError = err;
            logger.debug("The executor failed", err);
            if (dateNow() >= timesAt) {
                throw err;
            }
        } finally{
            if (!result) {
                logger.debug("Waiting because the result is", result);
                await delay(intervalInMs, options);
            }
        }
    }
    if (dateNow() >= timesAt) {
        if (lastError) {
            throw lastError;
        }
        throw new _timeoutError.default("Timed out till specified callable returns true");
    }
    return result;
}
function waitTrue(callable, intervalInMs = 25, options) {
    return wait(callable, intervalInMs, options);
}
async function waitPass(callable, intervalInMs = 25, options) {
    let result;
    await wait(async ()=>{
        result = await callable();
        return true;
    }, intervalInMs, options);
    return result;
}
async function waitAny(callable, intervalInMs = 25, options) {
    let result;
    await waitTrue(async ()=>{
        result = await callable();
        if (result !== undefined) {
            return true;
        }
    }, intervalInMs, options);
    return result;
}
function expBackoffDelay(iteration, startDelay, maxDelay) {
    if (iteration === 0) {
        return 0;
    }
    return Math.min(startDelay * Math.pow(2, iteration - 1), maxDelay);
}
