// metaapi.cloud-sdk@29.3.3/dist/helpers/promises.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ConditionPromise: function() {
        return ConditionPromise;
    },
    HandlePromise: function() {
        return HandlePromise;
    }
});
const _helpers = /*#__PURE__*/ _interop_require_wildcard(require("./helpers"));
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
let HandlePromise = class HandlePromise {
    /**
   * Returns whether the promise has resolved
   * @returns {boolean} whether the promise has resolved
   */ get resolved() {
        return !!this._promise.resolved;
    }
    /**
   * Returns whether the promise has rejected
   * @returns {boolean} whether the promise has rejected
   */ get rejected() {
        return !!this._promise.rejected;
    }
    /**
   * Returns whether the promise has completed (either resolved or rejected)
   * @returns {boolean} whether the promise has completed (either resolved or rejected)
   */ get completed() {
        return !!this._promise.completed;
    }
    /**
   * Constructs instance
   * @param {Promise<T>} [promise] original promise to wrap into the handle
   */ constructor(promise){
        this._promise = _helpers.createHandlePromise();
        this.then = this._promise.then.bind(this._promise);
        this.catch = this._promise.catch.bind(this._promise);
        this.finally = this._promise.finally.bind(this._promise);
        this.resolve = this._promise.resolve.bind(this._promise);
        this.reject = this._promise.reject.bind(this._promise);
        this.timeout = this._promise.timeout.bind(this._promise);
        if (promise) {
            promise.then(this.resolve).catch(()=>{});
            promise.catch(this.reject);
        }
    }
};
let ConditionPromise = class ConditionPromise extends HandlePromise {
    /**
   * Calls the predicate to check if the promise should be resolved
   */ check() {
        if (this._predicate()) {
            this.resolve();
        }
    }
    /**
   * Constructs instance
   * @param {() => boolean} predicate function that returns true when the condition is reached
   */ constructor(predicate){
        super();
        this._predicate = predicate;
        this.check();
    }
};
