// metaapi.cloud-sdk@29.3.3/build/cjs/clients/domain.client.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return DomainClient;
    }
});
const _logger = /*#__PURE__*/ _interop_require_default(require("../logger"));
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
let DomainClient = class DomainClient {
    /**
   * Returns domain client domain
   * @returns {String} client domain
   */ get domain() {
        return this._domain;
    }
    /**
   * Returns domain client token
   * @returns {String} client token
   */ get token() {
        return this._token;
    }
    /**
   * Returns the API URL
   * @param {String} host REST API host
   * @param {String} region host region
   * @returns {String} API URL
   */ async getUrl(host, region) {
        await this._updateDomain();
        return `${host}.${region}.${this._urlCache.domain}`;
    }
    /**
   * Returns domain settings
   * @returns {DomainSettings} domain settings
   */ async getSettings() {
        await this._updateDomain();
        return {
            domain: this._urlCache.domain,
            hostname: this._urlCache.hostname
        };
    }
    async _updateDomain() {
        if (!this._urlCache.domain || this._urlCache.lastUpdated < Date.now() - 1000 * 60 * 10) {
            if (this._urlCache.requestPromise) {
                await this._urlCache.requestPromise;
            } else {
                let resolve, reject;
                this._urlCache.requestPromise = new Promise((res, rej)=>{
                    resolve = res, reject = rej;
                });
                let isCacheUpdated = false;
                while(!isCacheUpdated){
                    const opts = {
                        url: `https://mt-provisioning-api-v1.${this._domain}/users/current/servers/mt-client-api`,
                        method: "GET",
                        headers: {
                            "auth-token": this._token
                        },
                        json: true
                    };
                    try {
                        const urlSettings = await this._httpClient.request(opts, "_updateDomain");
                        this._urlCache = {
                            domain: urlSettings.domain,
                            hostname: urlSettings.hostname,
                            requestPromise: null,
                            lastUpdated: Date.now()
                        };
                        resolve();
                        isCacheUpdated = true;
                        this._retryIntervalInSeconds = 1;
                    } catch (err) {
                        this._logger.error("Failed to update domain settings cache", err);
                        this._retryIntervalInSeconds = Math.min(this._retryIntervalInSeconds * 2, 300);
                        await new Promise((res)=>setTimeout(res, this._retryIntervalInSeconds * 1000));
                    }
                }
            }
        }
    }
    /**
   * Constructs domain client instance
   * @param {HttpClient} httpClient HTTP client
   * @param {String} token authorization token
   * @param {String} domain domain to connect to, default is agiliumtrade.agiliumtrade.ai
   */ constructor(httpClient, token, domain = "agiliumtrade.agiliumtrade.ai"){
        _define_property(this, "_httpClient", void 0);
        _define_property(this, "_domain", void 0);
        _define_property(this, "_token", void 0);
        _define_property(this, "_urlCache", void 0);
        _define_property(this, "_retryIntervalInSeconds", void 0);
        _define_property(this, "_logger", void 0);
        this._httpClient = httpClient;
        this._domain = domain;
        this._token = token;
        this._urlCache = {
            domain: null,
            hostname: null,
            requestPromise: null,
            lastUpdated: 0
        };
        this._retryIntervalInSeconds = 1;
        this._logger = _logger.default.getLogger("DomainClient");
    }
};
