// metaapi.cloud-sdk@29.3.3/build/cjs/packages/sticky-sockets/common.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    destroyInternalClientSocket: function() {
        return destroyInternalClientSocket;
    },
    getClientSocketIoTransportName: function() {
        return getClientSocketIoTransportName;
    },
    getSocketStickySessionId: function() {
        return getSocketStickySessionId;
    },
    sendHistory: function() {
        return sendHistory;
    },
    supportsStickySocketProtocol: function() {
        return supportsStickySocketProtocol;
    }
});
function getSocketStickySessionId(socket) {
    return socket.handshake.query["stickySocketConnectionId"] || socket.id; // eslint-disable-line dot-notation
}
function supportsStickySocketProtocol(socket) {
    return !!socket.handshake.query["stickySocketConnectionId"]; // eslint-disable-line dot-notation
}
function getClientSocketIoTransportName(socket) {
    return socket.io.engine.transport.name;
}
function destroyInternalClientSocket(socket, err) {
    socket.io.engine.transport.ws._socket.destroy(err);
}
function sendHistory(socket, history, startIndex) {
    while(history.length && history.front().index < startIndex){
        history.shift();
    }
    for (let packet of history){
        packet.time = new Date();
        socket.emit(packet.event, {
            index: packet.index,
            data: packet.data
        });
    }
}
