// metaapi.cloud-sdk@29.3.3/build/cjs/helpers/test/assert.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    arrayMatchWithEqualLength: function() {
        return arrayMatchWithEqualLength;
    },
    callCountWithExactly: function() {
        return callCountWithExactly;
    },
    callCountWithMatch: function() {
        return callCountWithMatch;
    },
    callOrder: function() {
        return callOrder;
    }
});
const _testHelpers = /*#__PURE__*/ _interop_require_wildcard(require("./testHelpers"));
const _assert = require("assert");
const _util = /*#__PURE__*/ _interop_require_default(require("util"));
require("should");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function callCountWithMatch(spy, calls, ...args) {
    if (!spy.isSinonProxy) {
        throw new _assert.AssertionError({
            message: "Given function is not a sinon spy"
        });
    }
    let matchedCalls = _testHelpers.getCallsWithMatch(spy, ...args);
    if (matchedCalls.length !== calls) {
        throw Object.assign(new _assert.AssertionError({
            message: `Spy expected to be called ${calls} times with specified arguments but was called ` + `${matchedCalls.length} times with them of ${spy.callCount} total calls`
        }), {
            callArgs: _util.default.inspect(spy.args, {
                depth: 3
            })
        });
    }
}
function callCountWithExactly(spy, calls, ...args) {
    if (!spy.isSinonProxy) {
        throw new _assert.AssertionError({
            message: "Given function is not a sinon spy"
        });
    }
    let matchedCalls = _testHelpers.getCallsWithExactly(spy, ...args);
    matchedCalls.length.should.equal(calls, `Spy expected to be called ${calls} times with specified arguments ` + `but was called ${matchedCalls.length} times with them of ${spy.callCount} total calls`);
}
function arrayMatchWithEqualLength(actualArray, expectedArray) {
    actualArray.should.match(expectedArray);
    actualArray.length.should.equal(expectedArray.length);
}
function callOrder(spyCalls) {
    let callIds = spyCalls.map((call, index)=>{
        const label = call.label || call.spy.name || "default";
        let sinonCall = call.spy.getCall(call.call);
        if (!sinonCall) {
            throw new _assert.AssertionError({
                message: `Call ${index} (${label}) does not exist`
            });
        }
        return {
            id: sinonCall.callId,
            label
        };
    });
    let previousCallId;
    for(let index = 0; index < spyCalls.length; ++index){
        if (previousCallId !== undefined && callIds[index].id < previousCallId.id) {
            throw new _assert.AssertionError({
                message: `Wrong call order detected in call ids ${callIds.map((id)=>`\n${id.id} (${id.label})`)}`,
                expected: `Call id larger than the previous call ${previousCallId}`,
                actual: callIds[index]
            });
        }
        if (spyCalls[index].matchArgs) {
            try {
                spyCalls[index].spy.getCall(spyCalls[index].call).args.should.match(spyCalls[index].matchArgs);
            } catch (err) {
                throw Object.assign(new _assert.AssertionError({
                    message: `Call ${index} (${callIds[index].label}) args do not match`
                }), {
                    index,
                    label: callIds[index].label,
                    cause: err
                });
            }
        }
        previousCallId = callIds[index];
    }
}
