// metaapi.cloud-sdk@29.3.3/dist/packages/sticky-sockets/clientStickySocket.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _randomstring = /*#__PURE__*/ _interop_require_default(require("randomstring"));
const _longrunningasyncprocess = /*#__PURE__*/ _interop_require_wildcard(require("../long-running-async-process"));
const _stickySocketConnection = /*#__PURE__*/ _interop_require_default(require("./stickySocketConnection"));
const _errorHandler = /*#__PURE__*/ _interop_require_wildcard(require("../../clients/errorHandler"));
const _commonutils = require("./common.utils");
const _time = require("../../helpers/convert/time");
const _eventEmitter = /*#__PURE__*/ _interop_require_default(require("../../tools/eventEmitter"));
const _logger = /*#__PURE__*/ _interop_require_default(require("../../logger"));
const _queue = /*#__PURE__*/ _interop_require_default(require("../../tools/queue"));
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
/**
 * Client sticky socket
 */ let ClientStickySocket = class ClientStickySocket extends _eventEmitter.default {
    /**
   * Returns socket ID
   * @returns socket ID
   */ get id() {
        return this._id;
    }
    /**
   * Returns current native socket
   * @returns socket.io socket
   * @internal intended for tests only
   */ get socket() {
        var _this__pool_getProcess;
        return (_this__pool_getProcess = this._pool.getProcess("connection")) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.socket;
    }
    /**
   * Returns current transport name, using socket.io internals
   * @returns transport name, e.g. "websocket"
   * @internal intended for tests only
   */ get transportName() {
        var _this__pool_getProcess;
        return (_this__pool_getProcess = this._pool.getProcess("connection")) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.transportName;
    }
    /**
   * Returns first present packet index in last emit history
   * @returns first history index
   * @internal intended for tests only
   */ get firstHistoryIndex() {
        var _this__pool_getProcess;
        return (_this__pool_getProcess = this._pool.getProcess("connection")) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.firstHistoryIndex;
    }
    /**
   * Returns whether socket connected
   * @returns is connected
   */ get connected() {
        return this._connected;
    }
    /**
   * @inheritdoc
   * @remarks Socket events `ClientStickySockets.INTERNAL_EVENTS` cannot be subscribed to
   * @throws `ValidationError` if attempting to subscribe to an internal event
   */ on(event, callback) {
        if (ClientStickySocket.INTERNAL_EVENTS.includes(event)) {
            throw new _errorHandler.ValidationError("Cannot subscribe to an internal event");
        }
        super.on(event, callback);
    }
    /**
   * Emits a data event. All events are buffered if no connection
   * @param event The event that we're emitting
   * @param args Optional arguments to send with the event
   */ send(event, ...args) {
        var _this__pool_getProcess;
        (_this__pool_getProcess = this._pool.getProcess("connection")) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.send(event, ...args);
    }
    /**
   * Disconnects socket
   * @param err error if disconnecting with error
   * @returns promise resolving when disconnected
   */ async disconnect(err) {
        clearTimeout(this._stickyConnectionWaitTimeout);
        let wasStopped = this._stopped;
        this._stopped = true;
        await this._pool.stop();
        if (!wasStopped) {
            this._connected = false;
            this.emit("disconnect", err);
        }
    }
    /**
   * Destroys underlying socket, using socket.io internals. The socket must have `websocket` transport
   * @param err optional error
   * @internal intended for tests only, e.g. to test connection loss
   */ destroyInternalSocket(err) {
        (0, _commonutils.destroyInternalClientSocket)(this._pool.getProcess("connection").socket, err);
    }
    /**
   * Removes expired emit history
   * @internal intended for tests only
   */ removeExpiredEmitHistory() {
        var _this__pool_getProcess;
        (_this__pool_getProcess = this._pool.getProcess("connection")) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.removeExpiredEmitHistory();
    }
    /**
   * Constructs instance
   * @param url URL
   * @param options options
   */ constructor(url, options){
        super();
        _define_property(this, "_logger", _logger.default.getLogger("ClientStickySocket"));
        _define_property(this, "_pool", void 0);
        _define_property(this, "_stickyConnectionWaitTimeout", void 0);
        _define_property(this, "_id", _randomstring.default.generate(16));
        _define_property(this, "_label", void 0);
        _define_property(this, "_stopped", false);
        _define_property(this, "_connected", false);
        this._label = `${(options === null || options === void 0 ? void 0 : options.label) || "connection"}:${this._id}`;
        var _options_stickyConnectionTtlInSeconds;
        const stickyConnectionTtlInMs = 1000 * ((_options_stickyConnectionTtlInSeconds = options === null || options === void 0 ? void 0 : options.stickyConnectionTtlInSeconds) !== null && _options_stickyConnectionTtlInSeconds !== void 0 ? _options_stickyConnectionTtlInSeconds : 10);
        let internalEvents = new _eventEmitter.default();
        internalEvents.once("connect", ()=>{
            this._connected = true;
            this.emit("connect");
        });
        internalEvents.on("connect", ()=>clearTimeout(this._stickyConnectionWaitTimeout));
        internalEvents.on("fail", (err)=>{
            if (options === null || options === void 0 ? void 0 : options.useNativeSocketIoServer) {
                this.disconnect(err);
                return;
            }
            if (!this._stopped) {
                this._stickyConnectionWaitTimeout = setTimeout(()=>{
                    this._logger.warn(`${this._label}: failed to restore sticky connection`);
                    this.disconnect(err);
                }, stickyConnectionTtlInMs);
            }
        });
        this._pool = new _longrunningasyncprocess.RootPool(_stickySocketConnection.default, {
            dependencies: [
                this,
                internalEvents
            ],
            label: `clientStickySocket:${this._label}`
        });
        var _options_emitHistoryTtlInSeconds, _options_reconnectionDelayInMs, _options_reconnectionDelayMaxInMs, _options_reconnectionRandomizationFactor;
        this._pool.scheduleProcess("connection", {
            args: [
                url,
                this._id,
                {
                    startCount: 0,
                    emitHistory: new _queue.default(),
                    lastSentIndex: -1,
                    lastReceivedIndex: -1
                },
                {
                    label: this._label,
                    connection: options === null || options === void 0 ? void 0 : options.connection,
                    useNativeSocketIoServer: options === null || options === void 0 ? void 0 : options.useNativeSocketIoServer,
                    emitHistoryTtlInSeconds: (_options_emitHistoryTtlInSeconds = options === null || options === void 0 ? void 0 : options.emitHistoryTtlInSeconds) !== null && _options_emitHistoryTtlInSeconds !== void 0 ? _options_emitHistoryTtlInSeconds : (0, _time.msToSeconds)(stickyConnectionTtlInMs)
                }
            ],
            failoverThrottleDelay: {
                mode: "exponential",
                resetDelayInMs: 0,
                minDelayInMs: (_options_reconnectionDelayInMs = options === null || options === void 0 ? void 0 : options.reconnectionDelayInMs) !== null && _options_reconnectionDelayInMs !== void 0 ? _options_reconnectionDelayInMs : 1000,
                maxDelayInMs: (_options_reconnectionDelayMaxInMs = options === null || options === void 0 ? void 0 : options.reconnectionDelayMaxInMs) !== null && _options_reconnectionDelayMaxInMs !== void 0 ? _options_reconnectionDelayMaxInMs : 10000,
                randomizationFactor: (_options_reconnectionRandomizationFactor = options === null || options === void 0 ? void 0 : options.reconnectionRandomizationFactor) !== null && _options_reconnectionRandomizationFactor !== void 0 ? _options_reconnectionRandomizationFactor : 0.5
            }
        });
    }
};
(function(ClientStickySocket) {
    /** Connection events */ ClientStickySocket.CONNECTION_EVENTS = _stickySocketConnection.default.CONNECTION_EVENTS;
    /** Internal events */ ClientStickySocket.INTERNAL_EVENTS = _stickySocketConnection.default.INTERNAL_EVENTS;
})(ClientStickySocket || (ClientStickySocket = {}));
const _default = ClientStickySocket;
