// metaapi.cloud-sdk@29.3.3/dist/packages/sticky-sockets/stickySocketConnection.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _longrunningasyncprocess = /*#__PURE__*/ _interop_require_wildcard(require("../long-running-async-process"));
const _socketioclient = /*#__PURE__*/ _interop_require_default(require("socket.io-client"));
const _socketiowildcard = /*#__PURE__*/ _interop_require_default(require("../socket.io-wildcard"));
const _commontypes = require("./common.types");
const _commonutils = require("./common.utils");
const _logger = /*#__PURE__*/ _interop_require_default(require("../../logger"));
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
/**
 * A socket connection
 */ let StickySocketConnection = class StickySocketConnection extends _longrunningasyncprocess.RootProcess {
    /**
   * Returns native socket
   * @returns socket.io socket
   */ get socket() {
        return this._socket;
    }
    /**
   * Returns current transport name, using socket.io internals
   * @returns transport name, e.g. "websocket"
   */ get transportName() {
        return (0, _commonutils.getClientSocketIoTransportName)(this._socket);
    }
    /**
   * Returns first present packet index in last emit history
   * @returns first history index
   * @internal intended for tests only
   */ get firstHistoryIndex() {
        var _this__sharedState_emitHistory_front;
        return (_this__sharedState_emitHistory_front = this._sharedState.emitHistory.front()) === null || _this__sharedState_emitHistory_front === void 0 ? void 0 : _this__sharedState_emitHistory_front.index;
    }
    /**
   * @inheritdoc
   * @param client parent client
   * @param internalEvents internal event emitter
   */ inject(client, internalEvents) {
        this._client = client;
        this._internalEvents = internalEvents;
    }
    /**
   * @inheritdoc
   * @param url url to connect to
   * @param sessionId session ID
   * @param sharedState shared state
   * @param options additiona options
   */ initialize(url, sessionId, sharedState, options) {
        var _options_connection;
        this._options = options;
        this._label = (options === null || options === void 0 ? void 0 : options.label) || "default";
        this._sharedState = sharedState;
        this._sharedState.startCount++;
        let query = {
            ...options === null || options === void 0 ? void 0 : (_options_connection = options.connection) === null || _options_connection === void 0 ? void 0 : _options_connection.query,
            stickySocketConnectionId: sessionId
        };
        if (sharedState.startCount > 1) {
            var _this__sharedState_emitHistory_front;
            let request = {
                lastReceivedIndex: sharedState.lastReceivedIndex,
                lastSentIndex: sharedState.lastSentIndex,
                firstHistoryIndex: (_this__sharedState_emitHistory_front = this._sharedState.emitHistory.front()) === null || _this__sharedState_emitHistory_front === void 0 ? void 0 : _this__sharedState_emitHistory_front.index,
                sessionId
            };
            this._logger.debug(`${this._label}: restoring session`, JSON.stringify(request));
            query.restoreStickyConnection = encodeURIComponent(JSON.stringify(request));
        }
        this._socket = (0, _socketioclient.default)(url, {
            ...options === null || options === void 0 ? void 0 : options.connection,
            reconnection: false,
            autoConnect: false,
            query
        });
        (0, _socketiowildcard.default)(_socketioclient.default.Manager)(this._socket);
        this._socket.on("*", (packet)=>{
            if (packet.type === _commontypes.EVENT || packet.type === _commontypes.BINARY_EVENT) {
                var _this__options;
                let payload = packet.data[1];
                if ((_this__options = this._options) === null || _this__options === void 0 ? void 0 : _this__options.useNativeSocketIoServer) {
                    this._client.emit(packet.data[0], ...packet.data.slice(1));
                    return;
                }
                if ("index" in payload) {
                    this._logger.trace(()=>`${this._label}: received packet ` + JSON.stringify({
                            index: payload.index
                        }));
                    var _this__sharedState_lastReceivedIndex;
                    this._sharedState.lastReceivedIndex = Math.max((_this__sharedState_lastReceivedIndex = this._sharedState.lastReceivedIndex) !== null && _this__sharedState_lastReceivedIndex !== void 0 ? _this__sharedState_lastReceivedIndex : -1, payload.index);
                    this._client.emit(packet.data[0], ...payload.data);
                }
            }
        });
    }
    /**
   * Emits a data event. All events are buffered if no connection
   * @param event The event that we're emitting
   * @param args Optional arguments to send with the event
   */ send(event, ...args) {
        var _this__options;
        if ((_this__options = this._options) === null || _this__options === void 0 ? void 0 : _this__options.useNativeSocketIoServer) {
            this._socket.emit(event, ...args);
            return;
        }
        let packet = {
            index: ++this._sharedState.lastSentIndex,
            event,
            data: args,
            time: new Date()
        };
        this._sharedState.emitHistory.push(packet);
        this._socket.emit(event, {
            index: packet.index,
            data: packet.data
        });
    }
    /**
   * @inheritdoc
   */ async start(stopPromise) {
        try {
            let connectPromise = this._connect(stopPromise);
            let restorePromise = this._restoreConnectionIfNeeded(stopPromise);
            let [_, restored] = await Promise.all([
                connectPromise,
                restorePromise
            ]);
            if (restored) {
                this._logger.info(`${this._label}: restored connection session`);
            }
        } catch (err) {
            this._logger.warn(`${this._label}: failed to connect`, err);
            if (this._sharedState.startCount === 1 || err instanceof RestoreRejectError) {
                this._client.disconnect(err);
            }
            throw new _longrunningasyncprocess.ControlSignal({
                action: "failover",
                severity: "info"
            });
        }
    }
    _connect(stopPromise) {
        this._socket.connect();
        return new Promise((resolve, reject)=>{
            stopPromise.then(()=>reject(new Error("Stopped during connection")));
            this._setSocketStageListener("connect", ()=>{
                // Emitting connect immediately to guarantee no data events will be received until connect event emitted
                this._logger.debug(`${this._label}: internal socket connected`);
                this._internalEvents.emit("connect");
                resolve();
            });
            this._setSocketStageListener("disconnect", (reason)=>{
                this._disconnectReason = reason;
                reject(new Error(`Disconnected when connecting due to ${reason}`));
            });
            this._setSocketStageListener("error", (err)=>reject(err));
            this._setSocketStageListener("connect_error", (err)=>reject(err));
            this._setSocketStageListener("connect_timeout", (err)=>reject(err));
        });
    }
    async _restoreConnectionIfNeeded(stopPromise) {
        if (this._sharedState.startCount === 1) {
            return false;
        }
        let sendSinceIndex = await new Promise((resolve, reject)=>{
            stopPromise.then(()=>reject(new Error("Stopped during restoring connection")));
            this._setSocketStageListener(_commontypes.RESTORE_CONNECTION_EVENT, (event)=>{
                event.restored ? resolve(event.sendSinceIndex) : reject(new RestoreRejectError("Cannot restore connection session"));
            });
            this._setSocketStageListener("disconnect", (reason)=>{
                reject(new Error(`Disconnected when restoring connection due to ${reason}`));
            });
            this._setSocketStageListener("error", (err)=>reject(err));
            this._setSocketStageListener("connect_error", (err)=>reject(err));
            this._setSocketStageListener("connect_timeout", (err)=>reject(err));
        });
        this._logger.debug(`${this._label}: sending history since ${sendSinceIndex} packet index`);
        (0, _commonutils.sendHistory)(this._socket, this._sharedState.emitHistory, sendSinceIndex);
        return true;
    }
    /**
   * @inheritdoc
   */ async run(stopPromise) {
        var _this__options;
        var _this__options_emitHistoryTtlInSeconds;
        const historyBufferTtlInMs = 1000 * ((_this__options_emitHistoryTtlInSeconds = (_this__options = this._options) === null || _this__options === void 0 ? void 0 : _this__options.emitHistoryTtlInSeconds) !== null && _this__options_emitHistoryTtlInSeconds !== void 0 ? _this__options_emitHistoryTtlInSeconds : 10);
        let clearOldHistoryInterval = setInterval(()=>this.removeExpiredEmitHistory(), historyBufferTtlInMs);
        try {
            if (this._disconnectReason) {
                if (!this._tryDisconnectClientGracefully(this._disconnectReason)) {
                    throw new Error(`Disconnected due to ${this._disconnectReason}`);
                }
                return;
            }
            this._removePrevStageListeners();
            await Promise.race([
                stopPromise,
                new Promise((resolve, reject)=>{
                    this._setSocketStageListener("disconnect", (reason)=>{
                        if (!this._tryDisconnectClientGracefully(reason)) {
                            reject(new Error(`Disconnected due to ${reason}`));
                        }
                    });
                    this._setSocketStageListener("error", reject);
                })
            ]);
        } catch (err) {
            this._logger.warn(`${this._label}: lost connection`, err);
            this._internalEvents.emit("fail", err);
            throw new _longrunningasyncprocess.ControlSignal({
                action: "failover",
                severity: "info"
            });
        } finally{
            this._removePrevStageListeners();
            clearInterval(clearOldHistoryInterval);
        }
    }
    /**
   * Removes expired emit history
   */ removeExpiredEmitHistory() {
        var _this__options;
        var _this__options_emitHistoryTtlInSeconds;
        const historyBufferTtlInMs = 1000 * ((_this__options_emitHistoryTtlInSeconds = (_this__options = this._options) === null || _this__options === void 0 ? void 0 : _this__options.emitHistoryTtlInSeconds) !== null && _this__options_emitHistoryTtlInSeconds !== void 0 ? _this__options_emitHistoryTtlInSeconds : 10);
        while(this._sharedState.emitHistory.length){
            if (Date.now() - this._sharedState.emitHistory.front().time.getTime() > historyBufferTtlInMs) {
                this._sharedState.emitHistory.shift();
            } else {
                break;
            }
        }
    }
    _tryDisconnectClientGracefully(reason) {
        if (reason === "io client disconnect" || reason === "io server disconnect") {
            this._logger.info(`${this._label}: disconnecting client due to ${reason}`);
            this._client.disconnect();
            return true;
        }
        return false;
    }
    /**
   * @inheritdoc
   */ async stop() {
        this._socket.disconnect();
    }
    _setSocketStageListener(event, listener) {
        var _this__prevStageListeners, _event;
        (_this__prevStageListeners = this._prevStageListeners)[_event = event] || (_this__prevStageListeners[_event] = new Set());
        this._prevStageListeners[event].add(listener);
        this._socket.on(event, listener);
        return listener;
    }
    _removePrevStageListeners() {
        for (let [event, listeners] of Object.entries(this._prevStageListeners)){
            for (let listener of listeners){
                this._socket.off(event, listener);
            }
        }
    }
    constructor(...args){
        super(...args);
        _define_property(this, "_options", void 0);
        _define_property(this, "_socket", void 0);
        _define_property(this, "_client", void 0);
        _define_property(this, "_sharedState", void 0);
        _define_property(this, "_internalEvents", void 0);
        _define_property(this, "_disconnectReason", void 0);
        _define_property(this, "_logger", _logger.default.getLogger(StickySocketConnection.name));
        _define_property(this, "_prevStageListeners", {});
        _define_property(this, "_label", void 0);
    }
};
(function(StickySocketConnection) {
    /** Connection events */ StickySocketConnection.CONNECTION_EVENTS = [
        "connect",
        "disconnect"
    ];
    /** Sticky connection protocol events */ StickySocketConnection.PROTOCOL_EVENTS = [
        _commontypes.RESTORE_CONNECTION_EVENT
    ];
    /** Internal events */ StickySocketConnection.INTERNAL_EVENTS = [
        "error",
        "connect_error",
        "connect_timeout",
        "reconnect",
        "reconnect_attempt",
        "reconnecting",
        "reconnect_error",
        "reconnect_failed",
        "ping",
        "pong",
        ...StickySocketConnection.PROTOCOL_EVENTS
    ];
})(StickySocketConnection || (StickySocketConnection = {}));
const _default = StickySocketConnection;
/**
 * Error, meaning that the server rejected the restore request. E.g. server may have already
 * removed old emit history, required by the client, so the restoring cannot be done
 */ let RestoreRejectError = class RestoreRejectError extends Error {
};
