// metaapi.cloud-sdk@29.3.3/dist/packages/long-running-async-process/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Context: function() {
        return _processContext.default;
    },
    // ChildProcess, ChildScheduler, ChildPool, ChildContext,
    ControlSignal: function() {
        return _controlSignal.default;
    },
    Pool: function() {
        return _asyncProcessPool.default;
    },
    Process: function() {
        return _asyncProcess.default;
    },
    RootContext: function() {
        return _rootProcessContext.default;
    },
    RootPool: function() {
        return _rootProcessPool.default;
    },
    RootProcess: function() {
        return _rootProcess.default;
    },
    TimeoutError: function() {
        return _timeoutError.default;
    }
});
const _asyncProcessPool = /*#__PURE__*/ _interop_require_default(require("./lrap/abstract/asyncProcessPool"));
const _asyncProcess = /*#__PURE__*/ _interop_require_default(require("./lrap/abstract/asyncProcess"));
const _processContext = /*#__PURE__*/ _interop_require_default(require("./lrap/abstract/processContext"));
const _rootProcess = /*#__PURE__*/ _interop_require_default(require("./lrap/root/rootProcess"));
const _rootProcessPool = /*#__PURE__*/ _interop_require_default(require("./lrap/root/rootProcessPool"));
const _rootProcessContext = /*#__PURE__*/ _interop_require_default(require("./lrap/root/rootProcessContext"));
const _controlSignal = /*#__PURE__*/ _interop_require_default(require("./lrap/controlSignal"));
const _timeoutError = /*#__PURE__*/ _interop_require_default(require("../../clients/timeoutError"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
