// metaapi.cloud-sdk@29.3.3/build/cjs/riskManagement/clients/equityTracking/trackerEventListenerManager.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return TrackerEventListenerManager;
    }
});
const _randomstring = /*#__PURE__*/ _interop_require_default(require("randomstring"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
let TrackerEventListenerManager = class TrackerEventListenerManager {
    /**
   * Returns the dictionary of tracker event listeners
   * @returns {{[listenerId: string]: TrackerEventListener}} dictionary of tracker event listeners
   */ get trackerEventListeners() {
        return this._trackerEventListeners;
    }
    /**
   * Adds a tracker event listener
   * @param {TrackerEventListener} listener tracker event listener 
   * @param {String} [accountId] account id
   * @param {String} [trackerId] tracker id
   * @param {Number} [sequenceNumber] event sequence number
   * @returns {String} tracker event listener id
   */ addTrackerEventListener(listener, accountId, trackerId, sequenceNumber) {
        const listenerId = _randomstring.default.generate(10);
        this._trackerEventListeners[listenerId] = listener;
        this._startTrackerEventJob(listenerId, listener, accountId, trackerId, sequenceNumber);
        return listenerId;
    }
    /**
   * Removes tracker event listener by id
   * @param {String} listenerId listener id 
   */ removeTrackerEventListener(listenerId) {
        delete this._trackerEventListeners[listenerId];
    }
    async _startTrackerEventJob(listenerId, listener, accountId, trackerId, sequenceNumber) {
        let throttleTime = this._errorThrottleTime;
        while(this._trackerEventListeners[listenerId]){
            try {
                const packets = await this._domainClient.requestApi({
                    url: "/users/current/tracker-events/stream",
                    method: "GET",
                    headers: {
                        "auth-token": this._domainClient.token,
                        "api-version": "1"
                    },
                    params: {
                        previousSequenceNumber: sequenceNumber,
                        accountId,
                        trackerId,
                        limit: 1000
                    }
                }, true);
                for (let packet of packets){
                    await listener.onTrackerEvent(packet);
                }
                throttleTime = this._errorThrottleTime;
                if (this._trackerEventListeners[listenerId] && packets.length) {
                    sequenceNumber = packets.slice(-1)[0].sequenceNumber;
                }
            } catch (err) {
                listener.onError(err);
                await new Promise((res)=>setTimeout(res, throttleTime));
                throttleTime = Math.min(throttleTime * 2, 30000);
            }
        }
    }
    /**
   * Constructs tracker event listener manager instance
   * @param {DomainClient} domainClient domain client
   */ constructor(domainClient){
        this._domainClient = domainClient;
        this._trackerEventListeners = {};
        this._errorThrottleTime = 1000;
    }
};
