// metaapi.cloud-sdk@29.3.3/build/esm/helpers/convert/time.js
'use strict';
/**
 * Converts milliseconds to seconds
 * @param ms milliseconds
 * @returns seconds
 */ export function msToSeconds(ms) {
    return ms / 1000;
}
/**
 * Converts milliseconds to minutes
 * @param ms milliseconds
 * @returns minutes
 */ export function msToMinutes(ms) {
    return ms / (1000 * 60);
}
/**
 * Converts milliseconds to hours
 * @param ms millieconds
 * @returns hours
 */ export function msToHours(ms) {
    return ms / (1000 * 60 * 60);
}
/**
 * Converts seconds to milliseconds
 * @param seconds
 * @returns milliseconds
 */ export function secondsToMs(seconds) {
    return 1000 * seconds;
}
/**
 * Converts minutes to milliseconds
 * @param minutes minutes
 * @returns milliseconds
 */ export function minutesToMs(minutes) {
    return 1000 * 60 * minutes;
}
/**
 * Converts minutes to seconds
 * @param minutes minutes
 * @returns seconds
 */ export function minutesToSeconds(minutes) {
    return minutes * 60;
}
/**
 * Converts hours to milliseconds
 * @param hours hours
 * @returns milliseconds
 */ export function hoursToMs(hours) {
    return 1000 * 60 * 60 * hours;
}
/**
 * Converts hours to minutes
 * @param hours hours
 * @returns minutes
 */ export function hoursToMinutes(hours) {
    return 60 * hours;
}
