// metaapi.cloud-sdk@29.3.3/build/cjs/helpers/convert/time.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    hoursToMinutes: function() {
        return hoursToMinutes;
    },
    hoursToMs: function() {
        return hoursToMs;
    },
    minutesToMs: function() {
        return minutesToMs;
    },
    minutesToSeconds: function() {
        return minutesToSeconds;
    },
    msToHours: function() {
        return msToHours;
    },
    msToMinutes: function() {
        return msToMinutes;
    },
    msToSeconds: function() {
        return msToSeconds;
    },
    secondsToMs: function() {
        return secondsToMs;
    }
});
function msToSeconds(ms) {
    return ms / 1000;
}
function msToMinutes(ms) {
    return ms / (1000 * 60);
}
function msToHours(ms) {
    return ms / (1000 * 60 * 60);
}
function secondsToMs(seconds) {
    return 1000 * seconds;
}
function minutesToMs(minutes) {
    return 1000 * 60 * minutes;
}
function minutesToSeconds(minutes) {
    return minutes * 60;
}
function hoursToMs(hours) {
    return 1000 * 60 * 60 * hours;
}
function hoursToMinutes(hours) {
    return 60 * hours;
}
