// metaapi.cloud-metastats-sdk@4.1.1/build/esm/clients/metaStats.client.js
'use strict';
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
export * from './metaStats.client.schemas';
let MetaStatsClient = class MetaStatsClient {
    /**
   * Returns metrics of MetaApi account. This API call is billable
   * https://metaapi.cloud/docs/metastats/restApi/api/calculateMetrics/
   * @param {String} accountId MetaApi account id
   * @param {Boolean} [includeOpenPositions] indicates whether open positions will be included
   * in the metrics, default false
   * @return {Metrics} account metrics
   */ getMetrics(accountId, includeOpenPositions = false) {
        var _this = this;
        return _async_to_generator(function*() {
            const getOpts = (host, id)=>({
                    url: host + `/users/current/accounts/${id}/metrics`,
                    method: 'GET',
                    headers: {
                        'auth-token': _this._domainClient.token
                    },
                    params: {
                        includeOpenPositions
                    },
                    json: true
                });
            const { metrics } = yield _this._domainClient.requestMetastats(getOpts, accountId);
            return metrics;
        })();
    }
    /**
   * Returns historical trades of MetaApi account
   * https://metaapi.cloud/docs/metastats/restApi/api/getHistoricalTrades/
   * @param {String} accountId MetaApi account id
   * @param {String} startTime start of time range, inclusive
   * @param {String} endTime end of time range, exclusive
   * @param {Boolean} [updateHistory] update historical trades before returning results. 
   * If set to true, the API call will be counted towards billable MetaStats API calls. 
   * If set to false, the API call is not billable. Default is true
   * @param {Number} [limit] pagination limit
   * @param {Number} [offset] pagination offset
   * @param {Number} [marketValue ] trade market value
   * @return {Array<Trade>} account historical trades
   */ getAccountTrades(accountId, startTime, endTime, updateHistory = true, limit = 1000, offset = 0) {
        var _this = this;
        return _async_to_generator(function*() {
            const getOpts = (host, id)=>({
                    url: host + `/users/current/accounts/${id}/historical-trades/${startTime}/${endTime}`,
                    method: 'GET',
                    headers: {
                        'auth-token': _this._domainClient.token
                    },
                    params: {
                        updateHistory,
                        limit,
                        offset
                    },
                    json: true
                });
            const { trades } = yield _this._domainClient.requestMetastats(getOpts, accountId);
            return trades;
        })();
    }
    /**
   * Returns open trades of MetaApi account. This API call is not billable
   * https://metaapi.cloud/docs/metastats/restApi/api/getOpenTrades/
   * @param {String} accountId MetaApi account id
   * @param {Number} [marketValue] trade market value
   * @return {Array<OpenTrade>} account historical trades
   */ getAccountOpenTrades(accountId) {
        var _this = this;
        return _async_to_generator(function*() {
            const getOpts = (host, id)=>({
                    url: host + `/users/current/accounts/${id}/open-trades`,
                    method: 'GET',
                    headers: {
                        'auth-token': _this._domainClient.token
                    },
                    json: true
                });
            const { openTrades } = yield _this._domainClient.requestMetastats(getOpts, accountId);
            return openTrades;
        })();
    }
    /**
   * Resets metrics and trade history for a trading account from MetaStats. The data will be downloaded from trading
   * account again when you calculate the MetaStats metrics next time. This API call is not billable
   * @param accountId MetaApi account id
   * @return promise resolving when removed
   */ resetMetrics(accountId) {
        var _this = this;
        return _async_to_generator(function*() {
            const getOpts = (host, id)=>({
                    url: host + `/users/current/accounts/${id}`,
                    method: 'DELETE',
                    headers: {
                        'auth-token': _this._domainClient.token
                    },
                    json: true
                });
            return _this._domainClient.requestMetastats(getOpts, accountId);
        })();
    }
    /**
   * Constructs MetaStats API client instance
   * @param {DomainClient} domainClient domain client
   */ constructor(domainClient){
        _define_property(this, "_domainClient", void 0);
        this._domainClient = domainClient;
    }
};
/**
 * metaapi.cloud MetaStats MetaTrader API client
 */ export { MetaStatsClient as default };
