// metaapi.cloud-sdk@29.3.3/build/esm/riskManagement/clients/equityTracking/equityBalanceStreamManager.js
'use strict';
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
import randomstring from 'randomstring';
import SynchronizationListener from '../../../clients/metaApi/synchronizationListener';
import LoggerManager from '../../../logger';
let EquityBalanceStreamManager = class EquityBalanceStreamManager {
    /**
   * Returns listeners for account
   * @param {String} accountId account id to return listeners for
   * @returns {{[listenerId: string]: EquityBalanceListener}} dictionary of account equity balance event listeners
   */ getAccountListeners(accountId) {
        if (!this._equityBalanceListeners[accountId]) {
            this._equityBalanceListeners[accountId] = {};
        }
        return this._equityBalanceListeners[accountId];
    }
    /**
   * Adds an equity balance event listener
   * @param {EquityBalanceListener} listener equity balance event listener
   * @param {String} accountId account id
   * @returns {Promise<string>} listener id
   */ // eslint-disable-next-line max-statements, complexity
    addEquityBalanceListener(listener, accountId) {
        var _this = this;
        return _async_to_generator(function*() {
            if (!_this._equityBalanceCaches[accountId]) {
                _this._equityBalanceCaches[accountId] = {
                    balance: null,
                    equity: null,
                    pendingInitalizationResolves: []
                };
            }
            const cache = _this._equityBalanceCaches[accountId];
            let connection = null;
            let retryIntervalInSeconds = _this._retryIntervalInSeconds;
            const getAccountListeners = ()=>_this.getAccountListeners(accountId);
            const pendingInitalizationResolves = _this._pendingInitalizationResolves;
            const synchronizationFlags = _this._accountSynchronizationFlags;
            const processEquityBalanceEvent = function() {
                var _ref = _async_to_generator(function*(equity, balance) {
                    if (_this._equityBalanceCaches[accountId]) {
                        if (equity !== cache.equity || balance && balance !== cache.balance) {
                            cache.equity = equity;
                            if (balance) {
                                cache.balance = balance;
                            }
                            if (cache.equity !== null && cache.balance !== null) {
                                Object.values(getAccountListeners()).forEach((accountListener)=>{
                                    accountListener.onEquityOrBalanceUpdated({
                                        equity: cache.equity,
                                        balance: cache.balance
                                    });
                                });
                            }
                        }
                    }
                });
                return function processEquityBalanceEvent(equity, balance) {
                    return _ref.apply(this, arguments);
                };
            }();
            let EquityBalanceStreamListener = class EquityBalanceStreamListener extends SynchronizationListener {
                onDealsSynchronized(instanceIndex, synchronizationId) {
                    var _this = this;
                    return _async_to_generator(function*() {
                        try {
                            if (!synchronizationFlags[accountId]) {
                                synchronizationFlags[accountId] = true;
                                Object.values(getAccountListeners()).forEach((accountListener)=>{
                                    accountListener.onConnected();
                                });
                            }
                            if (pendingInitalizationResolves[accountId]) {
                                pendingInitalizationResolves[accountId].forEach((resolve)=>resolve());
                                delete pendingInitalizationResolves[accountId];
                            }
                        } catch (err) {
                            Object.values(getAccountListeners()).forEach((accountListener)=>{
                                accountListener.onError(err);
                            });
                            _this._logger.error('Error processing onDealsSynchronized event for ' + `equity balance listener for account ${accountId}`, err);
                        }
                    })();
                }
                onDisconnected(instanceIndex) {
                    var _this = this;
                    return _async_to_generator(function*() {
                        try {
                            if (synchronizationFlags[accountId] && !connection.healthMonitor.healthStatus.synchronized) {
                                synchronizationFlags[accountId] = false;
                                Object.values(getAccountListeners()).forEach((accountListener)=>{
                                    accountListener.onDisconnected();
                                });
                            }
                        } catch (err) {
                            Object.values(getAccountListeners()).forEach((accountListener)=>{
                                accountListener.onError(err);
                            });
                            _this._logger.error('Error processing onDisconnected event for ' + `equity balance listener for account ${accountId}`, err);
                        }
                    })();
                }
                // eslint-disable-next-line complexity, max-statements
                onSymbolPriceUpdated(instanceIndex, price) {
                    var _this = this;
                    return _async_to_generator(function*() {
                        try {
                            if (pendingInitalizationResolves[accountId]) {
                                pendingInitalizationResolves[accountId].forEach((resolve)=>resolve());
                                delete pendingInitalizationResolves[accountId];
                            }
                        } catch (err) {
                            Object.values(getAccountListeners()).forEach((accountListener)=>{
                                accountListener.onError(err);
                            });
                            _this._logger.error('Error processing onSymbolPriceUpdated event for ' + `equity balance listener for account ${accountId}`, err);
                        }
                        // price data only contains equity
                        yield processEquityBalanceEvent(price.equity);
                    })();
                }
                onAccountInformationUpdated(instanceIndex, accountInformation) {
                    return _async_to_generator(function*() {
                        yield processEquityBalanceEvent(accountInformation.equity, accountInformation.balance);
                    })();
                }
            };
            const listenerId = randomstring.generate(10);
            const accountListeners = _this.getAccountListeners(accountId);
            accountListeners[listenerId] = listener;
            _this._accountsByListenerId[listenerId] = accountId;
            let isDeployed = false;
            const account = yield _this._metaApi.metatraderAccountApi.getAccount(accountId);
            while(!isDeployed){
                try {
                    yield account.waitDeployed();
                    isDeployed = true;
                } catch (err) {
                    listener.onError(err);
                    _this._logger.error(`Error wait for account ${accountId} to deploy, retrying`, err);
                    yield new Promise((res)=>setTimeout(res, retryIntervalInSeconds * 1000));
                    retryIntervalInSeconds = Math.min(retryIntervalInSeconds * 2, 300);
                }
            }
            if (!_this._equityBalanceConnections[accountId]) {
                retryIntervalInSeconds = _this._retryIntervalInSeconds;
                connection = account.getStreamingConnection();
                _this._equityBalanceConnections[accountId] = connection;
                const syncListener = new EquityBalanceStreamListener();
                connection.addSynchronizationListener(syncListener);
                let isSynchronized = false;
                while(!isSynchronized){
                    try {
                        yield connection.connect();
                        yield connection.waitSynchronized();
                        isSynchronized = true;
                    } catch (err) {
                        listener.onError(err);
                        _this._logger.error('Error configuring equity balance stream listener ' + `for account ${accountId}, retrying`, err);
                        yield new Promise((res)=>setTimeout(res, retryIntervalInSeconds * 1000));
                        retryIntervalInSeconds = Math.min(retryIntervalInSeconds * 2, 300);
                    }
                }
                retryIntervalInSeconds = _this._retryIntervalInSeconds;
            } else {
                connection = _this._equityBalanceConnections[accountId];
                if (!connection.healthMonitor.healthStatus.synchronized) {
                    if (!_this._pendingInitalizationResolves[accountId]) {
                        _this._pendingInitalizationResolves[accountId] = [];
                    }
                    let resolveInitialize;
                    let initializePromise = new Promise((res, rej)=>{
                        resolveInitialize = res;
                    });
                    _this._pendingInitalizationResolves[accountId].push(resolveInitialize);
                    yield initializePromise;
                }
            }
            return listenerId;
        })();
    }
    /**
   * Removes equity balance event listener by id
   * @param {String} listenerId listener id
   */ removeEquityBalanceListener(listenerId) {
        if (this._accountsByListenerId[listenerId]) {
            const accountId = this._accountsByListenerId[listenerId];
            delete this._accountSynchronizationFlags[accountId];
            delete this._accountsByListenerId[listenerId];
            if (this._equityBalanceListeners[accountId]) {
                delete this._equityBalanceListeners[accountId][listenerId];
            }
            if (this._equityBalanceConnections[accountId] && !Object.keys(this._equityBalanceListeners[accountId]).length) {
                this._equityBalanceConnections[accountId].close();
                delete this._equityBalanceConnections[accountId];
            }
        }
    }
    /**
   * Constructs equity balance event listener manager instance
   * @param {DomainClient} domainClient domain client
   * @param {MetaApi} metaApi metaApi SDK instance
   */ constructor(domainClient, metaApi){
        this._domainClient = domainClient;
        this._metaApi = metaApi;
        this._equityBalanceListeners = {};
        this._accountsByListenerId = {};
        this._equityBalanceConnections = {};
        this._equityBalanceCaches = {};
        this._accountSynchronizationFlags = {};
        this._pendingInitalizationResolves = {};
        this._retryIntervalInSeconds = 1;
        this._logger = LoggerManager.getLogger('EquityBalanceStreamManager');
    }
};
/**
 * Manager for handling equity balance event listeners
 */ export { EquityBalanceStreamManager as default };
