// metaapi.cloud-sdk@29.3.3/build/esm/clients/errorHandler.js
'use strict';
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
/**
 * Base class for API errors. Contains indication of HTTP status.
 */ const isNode = typeof process !== 'undefined' && process.release && process.release.name === 'node';
/**
 * Base class for API errors. Contains indication of HTTP status.
 */ export class ApiError extends Error {
    /**
   * Sets error code, used for i18n
   * @param {string} code error code for i18n
   */ set code(code) {
        this._code = code;
    }
    /**
   * Returns error code used for i18n
   * @return {string} error code
   */ get code() {
        return this._code;
    }
    /**
   * Set message arguments for i18n
   * @param {Array<Object>} args arguments for i18n
   */ set arguments(args) {
        this._args = args;
    }
    /**
   * Returns message arguments for i18n
   * @return {Array<Object>} message arguments for i18n
   */ get arguments() {
        return this._args;
    }
    /**
   * ApiError constructor
   * @param clazz error function
   * @param message error message
   * @param status HTTP status
   * @param url API request URL
   */ constructor(clazz, message, status, url){
        super(url ? message + '. Request URL: ' + url : message);
        /** HTTP status code */ _define_property(this, "status", void 0);
        /** API request URL */ _define_property(this, "url", void 0);
        _define_property(this, "_code", void 0);
        _define_property(this, "_args", void 0);
        this.name = 'ApiError';
        this.status = status;
        this.url = url;
        if (isNode && Error.captureStackTrace) {
            Error.captureStackTrace(this, clazz);
        }
    }
}
/**
 * Throwing this error results in 404 (Not Found) HTTP response code.
 */ export class NotFoundError extends ApiError {
    /**
   * Represents NotFoundError.
   * @param {string} message error message
   * @param {string} url API request URL
   */ constructor(message, url){
        super(NotFoundError, message, 404, url);
        this.name = 'NotFoundError';
    }
}
/**
 * Throwing this error results in 403 (Forbidden) HTTP response code.
 */ export class ForbiddenError extends ApiError {
    /**
   * Constructs forbidden error.
   * @param {string} message error message
   * @param {string} url API request URL
   */ constructor(message, url){
        super(ForbiddenError, message, 403, url);
        this.name = 'ForbiddenError';
    }
}
/**
 * Throwing this error results in 401 (Unauthorized) HTTP response code.
 */ export class UnauthorizedError extends ApiError {
    /**
   * Constructs unauthorized error.
   * @param {string} message error message
   * @param {string} url API request URL
   */ constructor(message, url){
        super(UnauthorizedError, message, 401, url);
        this.name = 'UnauthorizedError';
    }
}
/**
 * Represents validation error. Throwing this error results in 400 (Bad Request) HTTP response code.
 */ export class ValidationError extends ApiError {
    /**
   * Constructs validation error.
   * @param {string} message error message
   * @param {Object} details error data
   * @param {string} url API request URL
   */ constructor(message, details, url){
        super(ValidationError, message, 400, url);
        /** Validation error details */ _define_property(this, "details", void 0);
        this.name = 'ValidationError';
        this.details = details;
    }
}
/**
 * Represents unexpected error. Throwing this error results in 500 (Internal Error) HTTP response code.
 */ export class InternalError extends ApiError {
    /**
   * Constructs unexpected error.
   * @param {string} message error message
   * @param {string} url API request URL
   */ constructor(message, url){
        super(InternalError, message, 500, url);
        this.name = 'InternalError';
    }
}
/**
 * Represents conflict error. Throwing this error results in 409 (Conflict) HTTP response code.
 */ export class ConflictError extends ApiError {
    /**
   * Constructs conflict error.
   * @param {string} message error message
   * @param {string} url API request URL
   */ constructor(message, url){
        super(ConflictError, message, 409, url);
        this.name = 'ConflictError';
    }
}
/**
 * Represents too many requests error. Throwing this error results in 429 (Too Many Requests) HTTP response code.
 */ export class TooManyRequestsError extends ApiError {
    /**
   * Constructs too many requests error.
   * @param {string} message error message
   * @param {TooManyRequestsErrorMetadata} metadata error metadata
   */ constructor(message, metadata, url){
        super(TooManyRequestsError, message, 429, url);
        /** error metadata */ _define_property(this, "metadata", void 0);
        this.name = 'TooManyRequestsError';
        this.metadata = metadata;
    }
}
