// metaapi.cloud-sdk@29.3.3/build/esm/packages/sticky-sockets/clientStickySocket.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
import randomstring from 'randomstring';
import * as lrap from '../long-running-async-process';
import StickySocketConnection from './stickySocketConnection';
import * as errors from '../../clients/errorHandler';
import { destroyInternalClientSocket } from './common.utils';
import { msToSeconds } from '../../helpers/convert/time';
import EventEmitter from '../../tools/eventEmitter';
import LoggerManager from '../../logger';
import Queue from '../../tools/queue';
/**
 * Client sticky socket
 */ let ClientStickySocket = class ClientStickySocket extends EventEmitter {
    /**
   * Returns socket ID
   * @returns socket ID
   */ get id() {
        return this._id;
    }
    /**
   * Returns current native socket
   * @returns socket.io socket
   * @internal intended for tests only
   */ get socket() {
        var _this__pool_getProcess;
        return (_this__pool_getProcess = this._pool.getProcess('connection')) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.socket;
    }
    /**
   * Returns current transport name, using socket.io internals
   * @returns transport name, e.g. "websocket"
   * @internal intended for tests only
   */ get transportName() {
        var _this__pool_getProcess;
        return (_this__pool_getProcess = this._pool.getProcess('connection')) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.transportName;
    }
    /**
   * Returns first present packet index in last emit history
   * @returns first history index
   * @internal intended for tests only
   */ get firstHistoryIndex() {
        var _this__pool_getProcess;
        return (_this__pool_getProcess = this._pool.getProcess('connection')) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.firstHistoryIndex;
    }
    /**
   * Returns whether socket connected
   * @returns is connected
   */ get connected() {
        return this._connected;
    }
    /**
   * @inheritdoc
   * @remarks Socket events `ClientStickySockets.INTERNAL_EVENTS` cannot be subscribed to
   * @throws `ValidationError` if attempting to subscribe to an internal event
   */ on(event, callback) {
        if (ClientStickySocket.INTERNAL_EVENTS.includes(event)) {
            throw new errors.ValidationError('Cannot subscribe to an internal event');
        }
        super.on(event, callback);
    }
    /**
   * Emits a data event. All events are buffered if no connection
   * @param event The event that we're emitting
   * @param args Optional arguments to send with the event
   */ send(event, ...args) {
        var _this__pool_getProcess;
        (_this__pool_getProcess = this._pool.getProcess('connection')) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.send(event, ...args);
    }
    /**
   * Disconnects socket
   * @param err error if disconnecting with error
   * @returns promise resolving when disconnected
   */ disconnect(err) {
        var _this = this;
        return _async_to_generator(function*() {
            clearTimeout(_this._stickyConnectionWaitTimeout);
            let wasStopped = _this._stopped;
            _this._stopped = true;
            yield _this._pool.stop();
            if (!wasStopped) {
                _this._connected = false;
                _this.emit('disconnect', err);
            }
        })();
    }
    /**
   * Destroys underlying socket, using socket.io internals. The socket must have `websocket` transport
   * @param err optional error
   * @internal intended for tests only, e.g. to test connection loss
   */ destroyInternalSocket(err) {
        destroyInternalClientSocket(this._pool.getProcess('connection').socket, err);
    }
    /**
   * Removes expired emit history
   * @internal intended for tests only
   */ removeExpiredEmitHistory() {
        var _this__pool_getProcess;
        (_this__pool_getProcess = this._pool.getProcess('connection')) === null || _this__pool_getProcess === void 0 ? void 0 : _this__pool_getProcess.removeExpiredEmitHistory();
    }
    /**
   * Constructs instance
   * @param url URL
   * @param options options
   */ constructor(url, options){
        super();
        _define_property(this, "_logger", LoggerManager.getLogger('ClientStickySocket'));
        _define_property(this, "_pool", void 0);
        _define_property(this, "_stickyConnectionWaitTimeout", void 0);
        _define_property(this, "_id", randomstring.generate(16));
        _define_property(this, "_label", void 0);
        _define_property(this, "_stopped", false);
        _define_property(this, "_connected", false);
        this._label = `${(options === null || options === void 0 ? void 0 : options.label) || 'connection'}:${this._id}`;
        var _options_stickyConnectionTtlInSeconds;
        const stickyConnectionTtlInMs = 1000 * ((_options_stickyConnectionTtlInSeconds = options === null || options === void 0 ? void 0 : options.stickyConnectionTtlInSeconds) !== null && _options_stickyConnectionTtlInSeconds !== void 0 ? _options_stickyConnectionTtlInSeconds : 10);
        let internalEvents = new EventEmitter();
        internalEvents.once('connect', ()=>{
            this._connected = true;
            this.emit('connect');
        });
        internalEvents.on('connect', ()=>clearTimeout(this._stickyConnectionWaitTimeout));
        internalEvents.on('fail', (err)=>{
            if (options === null || options === void 0 ? void 0 : options.useNativeSocketIoServer) {
                this.disconnect(err);
                return;
            }
            if (!this._stopped) {
                this._stickyConnectionWaitTimeout = setTimeout(()=>{
                    this._logger.warn(`${this._label}: failed to restore sticky connection`);
                    this.disconnect(err);
                }, stickyConnectionTtlInMs);
            }
        });
        this._pool = new lrap.RootPool(StickySocketConnection, {
            dependencies: [
                this,
                internalEvents
            ],
            label: `clientStickySocket:${this._label}`
        });
        var _options_emitHistoryTtlInSeconds, _options_reconnectionDelayInMs, _options_reconnectionDelayMaxInMs, _options_reconnectionRandomizationFactor;
        this._pool.scheduleProcess('connection', {
            args: [
                url,
                this._id,
                {
                    startCount: 0,
                    emitHistory: new Queue(),
                    lastSentIndex: -1,
                    lastReceivedIndex: -1
                },
                {
                    label: this._label,
                    connection: options === null || options === void 0 ? void 0 : options.connection,
                    useNativeSocketIoServer: options === null || options === void 0 ? void 0 : options.useNativeSocketIoServer,
                    emitHistoryTtlInSeconds: (_options_emitHistoryTtlInSeconds = options === null || options === void 0 ? void 0 : options.emitHistoryTtlInSeconds) !== null && _options_emitHistoryTtlInSeconds !== void 0 ? _options_emitHistoryTtlInSeconds : msToSeconds(stickyConnectionTtlInMs)
                }
            ],
            failoverThrottleDelay: {
                mode: 'exponential',
                resetDelayInMs: 0,
                minDelayInMs: (_options_reconnectionDelayInMs = options === null || options === void 0 ? void 0 : options.reconnectionDelayInMs) !== null && _options_reconnectionDelayInMs !== void 0 ? _options_reconnectionDelayInMs : 1000,
                maxDelayInMs: (_options_reconnectionDelayMaxInMs = options === null || options === void 0 ? void 0 : options.reconnectionDelayMaxInMs) !== null && _options_reconnectionDelayMaxInMs !== void 0 ? _options_reconnectionDelayMaxInMs : 10000,
                randomizationFactor: (_options_reconnectionRandomizationFactor = options === null || options === void 0 ? void 0 : options.reconnectionRandomizationFactor) !== null && _options_reconnectionRandomizationFactor !== void 0 ? _options_reconnectionRandomizationFactor : 0.5
            }
        });
    }
};
(function(ClientStickySocket) {
    /** Connection events */ ClientStickySocket.CONNECTION_EVENTS = StickySocketConnection.CONNECTION_EVENTS;
    /** Internal events */ ClientStickySocket.INTERNAL_EVENTS = StickySocketConnection.INTERNAL_EVENTS;
})(ClientStickySocket || (ClientStickySocket = {}));
export default ClientStickySocket;
