// metaapi.cloud-sdk@29.3.3/build/esm/packages/sticky-sockets/stickySocketConnection.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
import * as lrap from '../long-running-async-process';
import ioClient from 'socket.io-client';
import socketWildcard from '../socket.io-wildcard';
import { BINARY_EVENT, EVENT, RESTORE_CONNECTION_EVENT } from './common.types';
import { getClientSocketIoTransportName, sendHistory } from './common.utils';
import LoggerManager from '../../logger';
/**
 * A socket connection
 */ let StickySocketConnection = class StickySocketConnection extends lrap.RootProcess {
    /**
   * Returns native socket
   * @returns socket.io socket
   */ get socket() {
        return this._socket;
    }
    /**
   * Returns current transport name, using socket.io internals
   * @returns transport name, e.g. "websocket"
   */ get transportName() {
        return getClientSocketIoTransportName(this._socket);
    }
    /**
   * Returns first present packet index in last emit history
   * @returns first history index
   * @internal intended for tests only
   */ get firstHistoryIndex() {
        var _this__sharedState_emitHistory_front;
        return (_this__sharedState_emitHistory_front = this._sharedState.emitHistory.front()) === null || _this__sharedState_emitHistory_front === void 0 ? void 0 : _this__sharedState_emitHistory_front.index;
    }
    /**
   * @inheritdoc
   * @param client parent client
   * @param internalEvents internal event emitter
   */ inject(client, internalEvents) {
        this._client = client;
        this._internalEvents = internalEvents;
    }
    /**
   * @inheritdoc
   * @param url url to connect to
   * @param sessionId session ID
   * @param sharedState shared state
   * @param options additiona options
   */ initialize(url, sessionId, sharedState, options) {
        var _options_connection;
        this._options = options;
        this._label = (options === null || options === void 0 ? void 0 : options.label) || 'default';
        this._sharedState = sharedState;
        this._sharedState.startCount++;
        let query = _object_spread_props(_object_spread({}, options === null || options === void 0 ? void 0 : (_options_connection = options.connection) === null || _options_connection === void 0 ? void 0 : _options_connection.query), {
            stickySocketConnectionId: sessionId
        });
        if (sharedState.startCount > 1) {
            var _this__sharedState_emitHistory_front;
            let request = {
                lastReceivedIndex: sharedState.lastReceivedIndex,
                lastSentIndex: sharedState.lastSentIndex,
                firstHistoryIndex: (_this__sharedState_emitHistory_front = this._sharedState.emitHistory.front()) === null || _this__sharedState_emitHistory_front === void 0 ? void 0 : _this__sharedState_emitHistory_front.index,
                sessionId
            };
            this._logger.debug(`${this._label}: restoring session`, JSON.stringify(request));
            query.restoreStickyConnection = encodeURIComponent(JSON.stringify(request));
        }
        this._socket = ioClient(url, _object_spread_props(_object_spread({}, options === null || options === void 0 ? void 0 : options.connection), {
            reconnection: false,
            autoConnect: false,
            query
        }));
        socketWildcard(ioClient.Manager)(this._socket);
        this._socket.on('*', (packet)=>{
            if (packet.type === EVENT || packet.type === BINARY_EVENT) {
                var _this__options;
                let payload = packet.data[1];
                if ((_this__options = this._options) === null || _this__options === void 0 ? void 0 : _this__options.useNativeSocketIoServer) {
                    this._client.emit(packet.data[0], ...packet.data.slice(1));
                    return;
                }
                if ('index' in payload) {
                    this._logger.trace(()=>`${this._label}: received packet ` + JSON.stringify({
                            index: payload.index
                        }));
                    var _this__sharedState_lastReceivedIndex;
                    this._sharedState.lastReceivedIndex = Math.max((_this__sharedState_lastReceivedIndex = this._sharedState.lastReceivedIndex) !== null && _this__sharedState_lastReceivedIndex !== void 0 ? _this__sharedState_lastReceivedIndex : -1, payload.index);
                    this._client.emit(packet.data[0], ...payload.data);
                }
            }
        });
    }
    /**
   * Emits a data event. All events are buffered if no connection
   * @param event The event that we're emitting
   * @param args Optional arguments to send with the event
   */ send(event, ...args) {
        var _this__options;
        if ((_this__options = this._options) === null || _this__options === void 0 ? void 0 : _this__options.useNativeSocketIoServer) {
            this._socket.emit(event, ...args);
            return;
        }
        let packet = {
            index: ++this._sharedState.lastSentIndex,
            event,
            data: args,
            time: new Date()
        };
        this._sharedState.emitHistory.push(packet);
        this._socket.emit(event, {
            index: packet.index,
            data: packet.data
        });
    }
    /**
   * @inheritdoc
   */ start(stopPromise) {
        var _this = this;
        return _async_to_generator(function*() {
            try {
                let connectPromise = _this._connect(stopPromise);
                let restorePromise = _this._restoreConnectionIfNeeded(stopPromise);
                let [_, restored] = yield Promise.all([
                    connectPromise,
                    restorePromise
                ]);
                if (restored) {
                    _this._logger.info(`${_this._label}: restored connection session`);
                }
            } catch (err) {
                _this._logger.warn(`${_this._label}: failed to connect`, err);
                if (_this._sharedState.startCount === 1 || err instanceof RestoreRejectError) {
                    _this._client.disconnect(err);
                }
                throw new lrap.ControlSignal({
                    action: 'failover',
                    severity: 'info'
                });
            }
        })();
    }
    _connect(stopPromise) {
        this._socket.connect();
        return new Promise((resolve, reject)=>{
            stopPromise.then(()=>reject(new Error('Stopped during connection')));
            this._setSocketStageListener('connect', ()=>{
                // Emitting connect immediately to guarantee no data events will be received until connect event emitted
                this._logger.debug(`${this._label}: internal socket connected`);
                this._internalEvents.emit('connect');
                resolve();
            });
            this._setSocketStageListener('disconnect', (reason)=>{
                this._disconnectReason = reason;
                reject(new Error(`Disconnected when connecting due to ${reason}`));
            });
            this._setSocketStageListener('error', (err)=>reject(err));
            this._setSocketStageListener('connect_error', (err)=>reject(err));
            this._setSocketStageListener('connect_timeout', (err)=>reject(err));
        });
    }
    _restoreConnectionIfNeeded(stopPromise) {
        var _this = this;
        return _async_to_generator(function*() {
            if (_this._sharedState.startCount === 1) {
                return false;
            }
            let sendSinceIndex = yield new Promise((resolve, reject)=>{
                stopPromise.then(()=>reject(new Error('Stopped during restoring connection')));
                _this._setSocketStageListener(RESTORE_CONNECTION_EVENT, (event)=>{
                    event.restored ? resolve(event.sendSinceIndex) : reject(new RestoreRejectError('Cannot restore connection session'));
                });
                _this._setSocketStageListener('disconnect', (reason)=>{
                    reject(new Error(`Disconnected when restoring connection due to ${reason}`));
                });
                _this._setSocketStageListener('error', (err)=>reject(err));
                _this._setSocketStageListener('connect_error', (err)=>reject(err));
                _this._setSocketStageListener('connect_timeout', (err)=>reject(err));
            });
            _this._logger.debug(`${_this._label}: sending history since ${sendSinceIndex} packet index`);
            sendHistory(_this._socket, _this._sharedState.emitHistory, sendSinceIndex);
            return true;
        })();
    }
    /**
   * @inheritdoc
   */ run(stopPromise) {
        var _this = this;
        return _async_to_generator(function*() {
            var _this__options;
            var _this__options_emitHistoryTtlInSeconds;
            const historyBufferTtlInMs = 1000 * ((_this__options_emitHistoryTtlInSeconds = (_this__options = _this._options) === null || _this__options === void 0 ? void 0 : _this__options.emitHistoryTtlInSeconds) !== null && _this__options_emitHistoryTtlInSeconds !== void 0 ? _this__options_emitHistoryTtlInSeconds : 10);
            let clearOldHistoryInterval = setInterval(()=>_this.removeExpiredEmitHistory(), historyBufferTtlInMs);
            try {
                if (_this._disconnectReason) {
                    if (!_this._tryDisconnectClientGracefully(_this._disconnectReason)) {
                        throw new Error(`Disconnected due to ${_this._disconnectReason}`);
                    }
                    return;
                }
                _this._removePrevStageListeners();
                yield Promise.race([
                    stopPromise,
                    new Promise((resolve, reject)=>{
                        _this._setSocketStageListener('disconnect', (reason)=>{
                            if (!_this._tryDisconnectClientGracefully(reason)) {
                                reject(new Error(`Disconnected due to ${reason}`));
                            }
                        });
                        _this._setSocketStageListener('error', reject);
                    })
                ]);
            } catch (err) {
                _this._logger.warn(`${_this._label}: lost connection`, err);
                _this._internalEvents.emit('fail', err);
                throw new lrap.ControlSignal({
                    action: 'failover',
                    severity: 'info'
                });
            } finally{
                _this._removePrevStageListeners();
                clearInterval(clearOldHistoryInterval);
            }
        })();
    }
    /**
   * Removes expired emit history
   */ removeExpiredEmitHistory() {
        var _this__options;
        var _this__options_emitHistoryTtlInSeconds;
        const historyBufferTtlInMs = 1000 * ((_this__options_emitHistoryTtlInSeconds = (_this__options = this._options) === null || _this__options === void 0 ? void 0 : _this__options.emitHistoryTtlInSeconds) !== null && _this__options_emitHistoryTtlInSeconds !== void 0 ? _this__options_emitHistoryTtlInSeconds : 10);
        while(this._sharedState.emitHistory.length){
            if (Date.now() - this._sharedState.emitHistory.front().time.getTime() > historyBufferTtlInMs) {
                this._sharedState.emitHistory.shift();
            } else {
                break;
            }
        }
    }
    _tryDisconnectClientGracefully(reason) {
        if (reason === 'io client disconnect' || reason === 'io server disconnect') {
            this._logger.info(`${this._label}: disconnecting client due to ${reason}`);
            this._client.disconnect();
            return true;
        }
        return false;
    }
    /**
   * @inheritdoc
   */ stop() {
        var _this = this;
        return _async_to_generator(function*() {
            _this._socket.disconnect();
        })();
    }
    _setSocketStageListener(event, listener) {
        var _this__prevStageListeners, _event;
        (_this__prevStageListeners = this._prevStageListeners)[_event = event] || (_this__prevStageListeners[_event] = new Set());
        this._prevStageListeners[event].add(listener);
        this._socket.on(event, listener);
        return listener;
    }
    _removePrevStageListeners() {
        for (let [event, listeners] of Object.entries(this._prevStageListeners)){
            for (let listener of listeners){
                this._socket.off(event, listener);
            }
        }
    }
    constructor(...args){
        super(...args);
        _define_property(this, "_options", void 0);
        _define_property(this, "_socket", void 0);
        _define_property(this, "_client", void 0);
        _define_property(this, "_sharedState", void 0);
        _define_property(this, "_internalEvents", void 0);
        _define_property(this, "_disconnectReason", void 0);
        _define_property(this, "_logger", LoggerManager.getLogger(StickySocketConnection.name));
        _define_property(this, "_prevStageListeners", {});
        _define_property(this, "_label", void 0);
    }
};
(function(StickySocketConnection) {
    /** Connection events */ StickySocketConnection.CONNECTION_EVENTS = [
        'connect',
        'disconnect'
    ];
    /** Sticky connection protocol events */ StickySocketConnection.PROTOCOL_EVENTS = [
        RESTORE_CONNECTION_EVENT
    ];
    /** Internal events */ StickySocketConnection.INTERNAL_EVENTS = [
        'error',
        'connect_error',
        'connect_timeout',
        'reconnect',
        'reconnect_attempt',
        'reconnecting',
        'reconnect_error',
        'reconnect_failed',
        'ping',
        'pong',
        ...StickySocketConnection.PROTOCOL_EVENTS
    ];
})(StickySocketConnection || (StickySocketConnection = {}));
export default StickySocketConnection;
/**
 * Error, meaning that the server rejected the restore request. E.g. server may have already
 * removed old emit history, required by the client, so the restoring cannot be done
 */ let RestoreRejectError = class RestoreRejectError extends Error {
};
