// metaapi.cloud-sdk@29.3.3/build/esm/helpers/helpers.js
'use strict';
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
import TimeoutError from '../clients/timeoutError';
import LoggerManager from '../logger';
import _ from 'lodash';
const realSetTimeout = setTimeout;
const realDateNow = Date.now.bind(Date);
/**
 * Creates a promise that can be used as a handle. It will not raise errors when rejected until it is explicitly
 * awaited or catch is set
 * @returns modified handle promise
 */ export function createHandlePromise() {
    let resolve, reject;
    let promise = new Promise((res, rej)=>{
        resolve = res;
        reject = rej;
    });
    promise.completed = false;
    promise.resolve = (result)=>{
        if (!promise.completed) {
            promise.completed = true;
            promise.resolved = true;
            promise.result = result;
            resolve(result);
        }
    };
    promise.reject = (err)=>{
        if (!promise.completed) {
            promise.completed = true;
            promise.rejected = true;
            promise.error = err;
            reject(err);
        }
    };
    promise.timeout = (milliseconds, errorMessage)=>{
        if (!promise.completed) {
            let timeout = setTimeout(()=>promise.reject(new TimeoutError(errorMessage)), milliseconds);
            promise.finally(()=>clearTimeout(timeout)).catch(()=>{});
        }
        return promise;
    };
    promise.catch(()=>{});
    return promise;
}
/**
 * Wraps a promise into a handle promise
 * @param promise native promise
 * @returns handle promise
 */ export function wrapHandlePromise(promise) {
    let result = createHandlePromise();
    promise.then(result.resolve).catch(()=>{});
    promise.catch(result.reject);
    return result;
}
/**
 * This function ensures that a promise is returned
 * @param call call
 * @returns promise
 */ export function ensurePromise(call) {
    return _ensurePromise.apply(this, arguments);
}
function _ensurePromise() {
    _ensurePromise = _async_to_generator(function*(call) {
        return call();
    });
    return _ensurePromise.apply(this, arguments);
}
/**
 * Waits specified delay
 * @param ms Milliseconds to wait
 * @param options Additional options
 * @return promise resolving when the delay has ended
 */ export function delay(ms, options) {
    let resolve;
    let timeout;
    let canceled = false;
    let result = new Promise((res)=>{
        timeout = (options === null || options === void 0 ? void 0 : options.ignoreSinonClock) ? realSetTimeout(res, ms) : setTimeout(res, ms);
        resolve = res;
    });
    Object.defineProperty(result, 'canceled', {
        get: ()=>canceled,
        enumerable: true,
        configurable: true
    });
    result.cancel = ()=>{
        canceled = true;
        clearTimeout(timeout);
        resolve();
    };
    return result;
}
/**
 * Assembles log4js config from logging level map
 * @param {Object} [config] log4js config
 * @param {String} [config.defaultLevel = 'INFO'] Default logging level
 * @param {Object} [config.levels] Logging levels
 * @return {Object} Log4js config
 */ export function assembleLog4jsConfig(config = {}) {
    let appenders = {
        console: {
            type: 'console'
        }
    };
    let categories = {
        default: {
            appenders: Object.keys(appenders),
            level: config.defaultLevel || 'INFO'
        }
    };
    Object.keys(config.levels || {}).forEach((category)=>{
        categories[category] = {
            appenders: Object.keys(appenders),
            level: config.levels[category]
        };
    });
    return {
        appenders,
        categories
    };
}
/**
 * Waits untill specified callable will pass successfully and return true. Uses log4js logger named `helpers.wait`
 * @param {() => boolean|Promise<boolean>} callable Callable to call until it returns true
 * @param {Number} [intervalInMs = 25] Interval in milliseconds between the checks
 * @param {WaitOptions & DelayOptions} [options] Additional wait options
 * @return {Promise} Promise resolving with callable return value when waited
 * @throws {Error|TimeoutError} Error from the callable or timeout error when timed out
 */ // eslint-disable-next-line complexity
export function wait(callable) {
    return _wait.apply(this, arguments);
}
function _wait() {
    _wait = _async_to_generator(function*(callable, intervalInMs = 25, options) {
        const logger = LoggerManager.getLogger('helpers.wait');
        if (typeof intervalInMs === 'object') {
            // for backward compatibility
            options = intervalInMs;
            intervalInMs = _.defaultTo(options.intervalInMs, 1000);
        }
        const dateNow = (options === null || options === void 0 ? void 0 : options.ignoreSinonClock) ? realDateNow : ()=>Date.now();
        let result = false, lastError;
        let timesAt = dateNow() + _.defaultTo(options === null || options === void 0 ? void 0 : options.timeoutInMs, 30000);
        while(!result && dateNow() < timesAt){
            try {
                result = yield callable();
            } catch (err) {
                lastError = err;
                logger.debug('The executor failed', err);
                if (dateNow() >= timesAt) {
                    throw err;
                }
            } finally{
                if (!result) {
                    logger.debug('Waiting because the result is', result);
                    yield delay(intervalInMs, options);
                }
            }
        }
        if (dateNow() >= timesAt) {
            if (lastError) {
                throw lastError;
            }
            throw new TimeoutError('Timed out till specified callable returns true');
        }
        return result;
    });
    return _wait.apply(this, arguments);
}
/**
 * Waits untill specified callable will pass successfully and return true. Uses log4js logger named `helpers.wait`
 * @param {() => boolean|Promise<boolean>} callable Callable to call until it returns true
 * @param {Number} [intervalInMs = 25] Interval in milliseconds between the checks
 * @param {WaitOptions & DelayOptions} [options] Additional wait options
 * @return {Promise} Promise resolving with callable return value when waited
 * @throws {Error|TimeoutError} Error from the callable or timeout error when timed out
 */ export function waitTrue(callable, intervalInMs = 25, options) {
    return wait(callable, intervalInMs, options);
}
/**
 * Waits untill specified callable will pass successfully. Uses log4js logger named `helpers.wait`
 * @param {() => boolean|Promise<boolean>} callable Callable to call
 * @param {Number} [intervalInMs = 25] Interval in milliseconds between the checks
 * @param {WaitOptions & DelayOptions} [options] Additional wait options
 * @return {Promise} Promise resolving with callable return value when waited
 * @throws {Error|TimeoutError} Error from the callable or timeout error when timed out
 */ export function waitPass(callable) {
    return _waitPass.apply(this, arguments);
}
function _waitPass() {
    _waitPass = _async_to_generator(function*(callable, intervalInMs = 25, options) {
        let result;
        yield wait(/*#__PURE__*/ _async_to_generator(function*() {
            result = yield callable();
            return true;
        }), intervalInMs, options);
        return result;
    });
    return _waitPass.apply(this, arguments);
}
/**
 * Waits untill specified callable successfully returns any non-undefined value. Uses log4js logger named `helpers.wait`
 * @param callable Callable to call
 * @param intervalInMs Interval in milliseconds between the checks
 * @param options Additional wait options
 * @return Promise resolving with callable return value when waited
 * @throws {Error|TimeoutError} Error from the callable or timeout error when timed out
 */ export function waitAny(callable) {
    return _waitAny.apply(this, arguments);
}
function _waitAny() {
    _waitAny = _async_to_generator(function*(callable, intervalInMs = 25, options) {
        let result;
        yield waitTrue(/*#__PURE__*/ _async_to_generator(function*() {
            result = yield callable();
            if (result !== undefined) {
                return true;
            }
        }), intervalInMs, options);
        return result;
    });
    return _waitAny.apply(this, arguments);
}
/**
 * Calculates exponential backoff delay. At the initial iteration, there is no delay. At the next iteration, the delay
 * is `startDelay`. Further, at the every next iteration the previous delay multiplies to 2
 * @param iteration current iteration, where 0 is initial iteration without delaying
 * @param startDelay start delay
 * @param maxDelay maximum delay
 */ export function expBackoffDelay(iteration, startDelay, maxDelay) {
    if (iteration === 0) {
        return 0;
    }
    return Math.min(startDelay * Math.pow(2, iteration - 1), maxDelay);
}
